superduperdb/misc/runnable/collection.py

```diff
@@ -0,0 +1,92 @@
+import dataclasses as dc
+import typing as t
+from functools import cached_property
+from queue import Empty, Queue
+
+from .runnable import Runnable
+from .thread import HasThread, print_err
+
+
+class HasRunnables(Runnable):
+    """Collect zero or more Runnable into one"""
+
+    runnables: t.Sequence[Runnable]
+
+    def start(self):
+        for r in self.runnables:
+            r.running.on_set.append(self._on_start)
+            r.stopped.on_set.append(self._on_stop)
+            r.start()
+
+    def stop(self):
+        self.running.clear()
+        for r in self.runnables:
+            r.stop()
+
+    def finish(self):
+        for r in self.runnables:
+            r.finish()
+
+    def join(self, timeout: t.Optional[float] = None):
+        for r in self.runnables:
+            r.join(timeout)
+
+    def _on_start(self):
+        if not self.running and all(r.running for r in self.runnables):
+            super().start()
+
+    def _on_stop(self):
+        if not self.stopped and all(r.stopped for r in self.runnables):
+            super().stop()
+
+
+@dc.dataclass
+class ThreadQueue(HasRunnables):
+    """A simple multi-producer, multi-consumer queue with one thread per consumer.
```

Comment:
Im confused here, if the class `_thread` method create a thread with a callback where you are consuming a queue.

```python
self.queue.get(...)
```
So, this is only intended to be used as multi-consumer right? 
producer part will still be executed by the user.
like queue.put(...)

Reply:
You can just hand the `queue.put` to any producer you like, insertions are thread-safe.